heron/ckptmgr/src/java/com/twitter/heron/ckptmgr/CheckpointManagerServer.java

```
@@ -0,0 +1,298 @@
+        + channel.socket().getRemoteSocketAddress());
+  }
+
+  @Override
+  public void onRequest(REQID rid, SocketChannel channel, Message request) {
+    if (request instanceof CheckpointManager.RegisterStMgrRequest) {
+      handleStMgrRegisterRequest(rid, channel, (CheckpointManager.RegisterStMgrRequest) request);
+    } else if (request instanceof CheckpointManager.RegisterTMasterRequest) {
+      handleTMasterRegisterRequest(rid, channel,
+                                   (CheckpointManager.RegisterTMasterRequest) request);
+    } else if (request instanceof CheckpointManager.SaveInstanceStateRequest) {
+      handleSaveInstanceStateRequest(
+          rid, channel, (CheckpointManager.SaveInstanceStateRequest) request);
+    } else if (request instanceof CheckpointManager.GetInstanceStateRequest) {
+      handleGetInstanceStateRequest(
+          rid, channel, (CheckpointManager.GetInstanceStateRequest) request);
+    } else if (request instanceof CheckpointManager.CleanStatefulCheckpointRequest) {
+      handleCleanStatefulCheckpointRequest(
+          rid, channel, (CheckpointManager.CleanStatefulCheckpointRequest) request);
+    } else {
+      LOG.severe("Unknown kind of request: " + request.getClass().getName());
+    }
+  }
+
+  protected void handleCleanStatefulCheckpointRequest(
+      REQID rid,
+      SocketChannel channel,
+      CheckpointManager.CleanStatefulCheckpointRequest request
+  ) {
+    LOG.info("Got a clean request from " + request.toString() + " host:port "
+        + channel.socket().getRemoteSocketAddress());
+
+    boolean deleteAll = request.hasCleanAllCheckpoints() && request.getCleanAllCheckpoints();
+    Common.StatusCode statusCode = Common.StatusCode.OK;
+
+    try {
+      checkpointsBackend.dispose(topologyName,
+                                 request.getOldestCheckpointPreserved(), deleteAll);
+      LOG.info("Dispose checkpoint successful");
+    } catch (StatefulStorageException e) {
```

If the exception runs all the way to slaveLooper runnable, that would actually kill the ckptmanager. Thus its kind of a fail fast approach.